presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -282,6 +284,18 @@ private void chooseDistribution(PartitioningHandle distribution)
                     partitioningHandle);
         }
 
+        private boolean isCompatibleSystemPartitioning(PartitioningHandle distribution)
+        {
+            ConnectorPartitioningHandle currentHandle = partitioningHandle.get().getConnectorHandle();
+            ConnectorPartitioningHandle distributionHandle = distribution.getConnectorHandle();
+            if ((currentHandle instanceof SystemPartitioningHandle) &&
+                    (distributionHandle instanceof SystemPartitioningHandle)) {
+                return ((SystemPartitioningHandle) currentHandle).getPartitioning() ==
```

Comment:
Ok, I see. The way this whole thing currently works is just the wrong abstraction. It's using the output partitioning to determine the "distribution" of the next fragment, but it's non-sensical to say something is "fixed round-robin distributed" or "fixed broadcast distributed". It does make sense to say "fixed distributed, based on some hash function (but of what data? input? output?)" or "fixed distributed based on some arbitrary but unknown function".